### DISCUSSION TITLE
How are exogenous variables handled in make_reduction?

### DISCUSSION THREAD

**Discussion by wregter** [alan-turing-institute/sktime#3332] [Q&A]
When I run the code below I see that make_reduction has created a LinearRegression model with 15 coefficients with window_length 5 and 2 exogenous variables. To me it seems that make_reduction is not only creating lag variables for my y but also my X. Is this also the way it is handled? How can I have control over this? For example, I would not like to use lags for holiday indicators in X or for static features for different time series when forecasting panel data. I have tried using WindowSummarizer inside make_reduction with window_length=None, but this threw me an error when adjusting the code below.

```
import numpy as np
import pandas as pd
from sktime.datasets import load_airline
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.base import ForecastingHorizon
from sklearn.linear_model import LinearRegression

y = load_airline()
X = pd.DataFrame({
        'X1': np.random.rand(y.shape[0]), 
        'X2': np.random.rand(y.shape[0])}
    ).set_index(y.index)

y_train, y_test, X_train, X_test = temporal_train_test_split(y, X, test_size=5)


forecaster = make_reduction(LinearRegression(), strategy='recursive', window_length=5)
fh = ForecastingHorizon(y_test.index, is_relative=False)
forecaster.fit(y_train, X=X_train, fh=fh)
print(forecaster.estimator_.coef_)
```

System:
    python: 3.8.13 (default, Mar 28 2022, 06:59:08) [MSC v.1916 64 bit (AMD64)]
executable: c:\Users\WREGTE16\Miniconda3\envs\forecasting\python.exe
   machine: Windows-10-10.0.19044-SP0

Python dependencies:
          pip: 22.1.2
   setuptools: 61.2.0
      sklearn: 1.1.2
       sktime: 0.13.2
  statsmodels: 0.13.2
        numpy: 1.22.4
        scipy: 1.8.1
       pandas: 1.4.3
   matplotlib: 3.5.3
       joblib: 1.1.0
        numba: 0.56.0
     pmdarima: None
      tsfresh: None

**Comment by fkiraly** [comment]
correct, this is known an unexpected behaviour. Reducers should not lag `X`.
Can't find the issue at the moment where this was first reported, but you are not the first one to.

Two relevant items:
* we're currently refactoring/reworking the reducers module. Issue and discussion here: https://github.com/alan-turing-institute/sktime/issues/3224, and a list of people who are involved. The problematic behaviour is referenced in the first bullet point.
* if you want to avoid the wonky behaviour, you could use `DirectReductionForecaster` in `sktime.forecasting.compose._reduce`. This is still experimental, but behaves as you would expect.

All feedback is appreciated, if you have suggestions, please put them in #3224!

**Reply by fkiraly** [reply]
Here's a PR for global/local reduction if you are interested. Also experimental, not in a release yet:
https://github.com/alan-turing-institute/sktime/pull/3327

**Comment by carlobindams** [comment]
Is that issue really solved? I'm using sktime v0.15 and that is still happening,.I was able to remove this behaviour only setting "pooling" = "global", but it looks like this slows down the predict method quite a lot! Is that fully fixed in later versions?

**Reply by fkiraly** [reply]
`sktime` is on 0.23.0 now, 0.15.0 is very old!
Could you give it a try with 0.23.0?

**Reply by carlobindams** [reply]
I'm sorry guys, the bug is still there even with version 23.0. so basically, to calculate only lags for y variable, I need to set make_reduction class to work with pooling="global". But to let you know, that slows down a lot performances when calling predict method.

**Comment by gkayode** [comment]
Hi @fkiraly. Do you know the order in which make_reduction creates this lags? In the [example](https://github.com/sktime/sktime/discussions/3332#discussion-4330428) above for instance. For the 15 coefficients, is it in the order [X1,X2,Y(t-1),X1(t-1),X2(t-1),Y(t-2)....]. The goal of this is to actually determine the coeff of the original feature X1, X2, so as to determine their feature importance.


I have a similar situation with xgboost as the base model with  window_length 17 and 15 exogenous variables. As you imagine, make_ reduction generated 272 features. However, i'm unsure the order in which this 272 feature is fed into the base model. All i want is to identify the feature importance for the original 15 variables.

**Reply by fkiraly** [reply]
@gkayode, this is a known issue with `make_reduction`. You could use `DirectReductionForecaster` which is an alternative implementation passing on the features, but it is slower.

**Reply by gkayode** [reply]
Thank you @fkiraly. Unrelated to this one issue i noticed is that i'm unable to successfully save and load model made with the TapNetRegressor. From errors ranging from `can't pickle weakref objects` to all others. I tried libraries like pickle, joblib and dill. All to no avail.
